data_structures/segment_tree.cpp

```diff
@@ -0,0 +1,128 @@
+/**
+ * @file
+ * @brief A data structure to quickly do operations on ranges: [Segment Tree](https://en.wikipedia.org/wiki/Segment_tree)
+ * @details
+ * Implementation of the segment tree data structre
+ *
+ * Can do point updates (updates the value of some position)
+ * and range queries, where it gives the value of some associative
+ * opperation done on a range
+ *
+ * Both of these operations take O(log N) time
+ * @author [Nishant Chatterjee](https://github.com/nishantc1527)
+ */
+
+#include <iostream>     // For IO operations
+#include <vector>       // For std::vector
+#include <algorithm>    // For std::min and std::max
+#include <cassert>      // For assert
+
+/*
+ * @namespace data_structures
+ */
+namespace data_structures {
+/**
+ * @brief class representation of the segment tree
+ * @param n The size of the tree
+ * @tparam T The type of the class that goes in the datastructure
+ */
+template <class T>
+class SegmentTree {
+private:
+    const T ID = 0;   ///< Comb(ID, x) = x
+    std::vector<T> t; ///< Vector to represent the tree
+    int size;

```

Comment:
Best to initialize it at zero to avoid problems.
```suggestion
    int size = 0;
```